reckon-core/src/main/groovy/org/ajoberstar/reckon/core/strategy/StagePreReleaseStrategy.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2015-2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ajoberstar.reckon.core.strategy;
+
+import com.github.zafarkhaja.semver.Version;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.ajoberstar.reckon.core.PreReleaseStrategy;
+import org.ajoberstar.reckon.core.VcsInventory;
+import org.ajoberstar.reckon.core.Versions;
+
+public final class StagePreReleaseStrategy implements PreReleaseStrategy {
+  public static final String FINAL_STAGE = "final";
+  private static final Pattern STAGE_REGEX = Pattern.compile("^(?<name>\\w+)(?:\\.(?<num>\\d+))?");
+
+  private final Set<String> stages;
+  private final String defaultStage;
+  private final Supplier<String> stageSupplier;
+
+  public StagePreReleaseStrategy(Set<String> stages, Supplier<String> stageSupplier) {
+    this.stages = stages;
+    this.defaultStage =
+        stages
+            .stream()
+            .sorted()
+            .findFirst()
+            .orElseThrow(() -> new IllegalArgumentException("No stages provided."));
+    this.stageSupplier = stageSupplier;
+  }
+
+  @Override
+  public Version reckonTargetVersion(VcsInventory inventory, Version targetNormal) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.ajoberstar.com/coding_rules#rule_key=squid%3AS3776)